src/lib.rs

```diff
@@ -63,6 +63,18 @@ pub enum Atom {
     // it might happen that it contains several leaves.
     DeleteBegin,
     DeleteEnd,
+    /// Scoped commands
+    // ScopedSoftline works together with the @open_scope and @end_scope query tags.
+    // To decide if a scoped softline must be expanded into a hardline, we look at
+    // the innermost scope earing the corresponding `scope_id`, that encompasses it.
+    // We expand the softline if that scope is multi-line.
+    // The `id` value is here for technical reasons, it allows tracking of the atom
+    // during post-processing.
```

Comment:
I had this question in mind when I was looking through the changes to `atom_collection.rs`, so I'm glad you mentioned it here :+1: Could you expand the explanation a bit, or include it in the post-processing function (I don't remember seeing it there)?

Reply:
Is it clearer with the new documentation for the `post_process_scopes` function?
```rust
    /// This function expands ScopedSoftline atoms depending on whether the context
    // containing them is multiline.
    // It does two passes over the atom collection: the first one associates each ScopedSoftline
    // to its scope, and decides what to replace them with when the scope ends.
    // The second pass applies the modifications to the atoms.
    fn post_process_scopes(&mut self) {
        type ScopeId = String;
        type LineIndex = usize;
        type ScopedNodeId = usize;
        // `opened_scopes` maintains stacks of opened scopes,
        // the line at which they started,
        // and the list of ScopedSoftline they contain.
        let mut opened_scopes: HashMap<&ScopeId, Vec<(LineIndex, Vec<&Atom>)>> = HashMap::new();
        // We can't process ScopedSoftlines in-place as we encounter them in the list of
        // atoms: we need to know when their encompassing scope ends to decide what to
        // replace them with. Instead of in-place modifications, we associate a replacement
        // atom to each ScopedSoftline atom (identified by their `id` field), then apply
        // the modifications in a second pass over the atoms.
        let mut modifications: HashMap<ScopedNodeId, Option<Atom>> = HashMap::new();
```

Reply:
LGTM :100: